x/application/keeper/prune_undelegations.go

```diff
@@ -0,0 +1,37 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	sessionkeeper "github.com/pokt-network/poktroll/x/session/keeper"
+)
+
+// undelegationRetentionSessions is the number of sessions for which undelegations
+// are retained before being pruned.
+// TODO_TECHDEBT: This should be a configuration parameter and potentially
+// a governance parameter.
+const undelegationRetentionSessions = 2
+
+// EndBlockerPruneUndelegations runs at the end of each block and prunes
+// undelegations that are older than the retention period.
+func (k Keeper) EndBlockerPruneUndelegations(ctx sdk.Context) error {
+	currentHeight := ctx.BlockHeight()
+	// Calculate the block height at which undelegations should be pruned
+	undelegationRetentionBlocks := sessionkeeper.GetSessionGracePeriodBlockCount() +
+		(sessionkeeper.NumBlocksPerSession * undelegationRetentionSessions)
+
+	// Iterate over all applications and prune undelegations that are older than
+	// the retention period.
+	for _, application := range k.GetAllApplications(ctx) {
+		for undelegationsHeight := range application.Undelegations {
+			if undelegationsHeight < uint64(currentHeight-undelegationRetentionBlocks) {
```

Comment:
Should this not be > ?

Reply:
I'm changing it, but just wanted to call out that the following is what helped me identify this potential bug:

1. Extra local variables
2. Better naming

Other than that, regardless of whether I'm right or wrong on this, this is why we need unit/integration (not E2E tests) DEDICATED to this.

Reply:
I still believe it is `<`. We are pruning undelegations that are **older** than `pruningBlockHeight` which means, if they have a **smaller** block height than `pruningBlockHeight` then they should be deleted.

Reply:
Okay, agreed. Let's change it to `if undelegationSessionEndHeight < pruningBlockHeight  then delete`, Thoughts?

Even though you are right, the fact that a test didn't fail for this still gives me "some" points though ;)